xarray/core/dataarray.py

```diff
@@ -484,6 +484,10 @@ def __setitem__(self, key, value):
         if isinstance(key, basestring):
             self.coords[key] = value
         else:
+            # Coordinates in key, value and self[key] should be consistent.
+            obj = self[key]
+            if isinstance(value, DataArray):
+                assert_coordinate_consistent(value, obj.coords)
```

Comment:
I think if you use `obj.coords.variables` here we can skip the awkward `getattr(coords[k], 'variable', coords[k])` above.